src/MongoDB.Driver/Linq/Processors/ParameterExpressionCollector.cs

```diff
@@ -0,0 +1,53 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to collect all parameters in the expression.
+    /// </summary>
+    internal sealed class ParameterExpressionCollector : ExpressionVisitor
+    {
+        private readonly List<string> _paramsCollection = new List<string>();
```

Comment:
Why are we using names? Was using the ParameterExpression directly and using referential equality not working? It feels like this could cause problems if a user shadows a name.

Reply:
Ok, I can do this, but what do you mean by this: 
> if a user shadows a name.

Reply:
Find(x => x.People.Any(x => x.Age > 23))

Note that x is shadowing (covering over) the parent x.

Reply:
I'm not sure, but I would expect we will have the syntax error in this case?
`Expression<Func<C, bool>> filter1 = x => x.S.Any(x=>x.D == "test")`
![image](https://user-images.githubusercontent.com/44870738/51191140-da981e80-18f4-11e9-923e-cceef2c47502.png)

Reply:
Cool... working in too many languages and forgetting what is allowed/disallowed. Is this a warning that is turned into an error because of our configuration? Or is this a real compilation error?

Reply:
a real compilation error